sinks/hawkular/driver.go

```diff
@@ -0,0 +1,236 @@
+type hawkularSink struct {
+	client  *metrics.Client
+	models  map[string]metrics.MetricDefinition // Model definitions
+	regLock sync.Mutex
+	reg     map[string]*metrics.MetricDefinition // Real definitions
+}
+
+// START: ExternalSink interface implementations
+
+func (self *hawkularSink) Register(mds []sink_api.MetricDescriptor) error {
+	self.regLock.Lock()
+	defer self.regLock.Unlock()
+
+	// Create model definitions based on the MetricDescriptors
+	for _, md := range mds {
+		hmd := self.descriptorToDefinition(&md)
+		self.models[md.Name] = hmd
+	}
+
+	// Fetch currently known metrics from Hawkular-Metrics and cache them
+	prev, err := self.client.Definitions(metrics.Gauge)
+	if err != nil {
+		return err
+	}
+
+	for _, p := range prev {
+		self.reg[p.Id] = p
+	}
+
+	return nil
+}
+
+// Transform the MetricDescriptor to a format used by Hawkular-Metrics
+func (self *hawkularSink) descriptorToDefinition(md *sink_api.MetricDescriptor) metrics.MetricDefinition {
+	tags := make(map[string]string)
+	// Postfix description tags with _description
+	for _, l := range md.Labels {
+		if len(l.Description) > 0 {
+			tags[l.Key+desc_tag] = l.Description
+		}
+	}
+
+	if len(md.Units.String()) > 0 {
+		tags[units_tag] = md.Units.String()
+	}
+	if len(md.Type.String()) > 0 {
+		tags[type_tag] = md.Type.String()
+	}
+
+	hmd := metrics.MetricDefinition{
+		Id:   md.Name,
+		Tags: tags,
+	}
+
+	return hmd
+}
+
+func (self *hawkularSink) groupName(p *sink_api.Point) string {
+	n := []string{p.Labels[sink_api.LabelContainerName], p.Name}
+	return strings.Join(n, sep_str)
+}
+
+func (self *hawkularSink) idName(p *sink_api.Point) string {
+	n := []string{p.Labels[sink_api.LabelContainerName], p.Labels[sink_api.LabelPodId], p.Name}
+	return strings.Join(n, sep_str)
+}
+
+// Check that metrics tags are defined on the Hawkular server and if not,
+// register the metric definition.
+func (self *hawkularSink) registerIfNecessary(t *sink_api.Timeseries) error {
+	key := self.idName(t.Point)
+
+	self.regLock.Lock()
+	defer self.regLock.Unlock()
+
+	// If found, check it matches the current stored definition (could be old info from
+	// the stored metrics cache for example)
+	if _, found := self.reg[key]; !found {
+		// Register the metric descriptor here..
+		if md, f := self.models[t.MetricDescriptor.Name]; f {
+			// Set tag values
+			for k, v := range t.Point.Labels {
+				md.Tags[k] = v
+			}
+
+			md.Tags[group_tag] = self.groupName(t.Point)
+
+			// Create metric
+			if err := self.client.UpdateTags(metrics.Gauge, t.MetricDescriptor.Name, md.Tags); err != nil {
+				// Log error and don't add this key to the lookup table
+				return err
+			}
+
+			// Add to the lookup table
+			self.reg[key] = &md
+			glog.Infof("Registered new metric definition: %s", key)
+		} else {
+			return fmt.Errorf("Could not find definition model with name %s", t.MetricDescriptor.Name)
+		}
+	}
+	// TODO Compare the definition tags and update if necessary? Quite expensive operation..
```

Comment:
Since this is a one time operation, I think we can disregard the expense. WDYT? Have updates would make user experience nicer. Very soon, we hope to lock down the current storage schema with appropriate versioning.